### DISCUSSION TITLE
Beginner struggling to improve mAP with multi-animal bottom-up ID (social interaction, heavy overlap, identity tracking)

### DISCUSSION THREAD

**Discussion by naaguilarr** [talmolab/sleap#2868] [Help!]
Hi everyone,

I'm a beginner with SLEAP and I'm trying to optimize my training configuration to improve my metrics, mainly mAP. Any guidance would be very welcome.

I'm analyzing social interaction between two animals that overlap heavily and need individual identity tracking, so I'm using the **multi-animal bottom-up ID** pipeline (`multi_class_bottomup`), which has worked better for me than the other pipelines I tried.

**Dataset:**
- 12 videos, 27,021 frames total
- 1,065 labeled frames
- Image quality is not great

**Skeleton (8 nodes):** nose, leftear, rightear, nape, center, tailbase, tailmiddle, tailtip
Should I consider adding more nodes (e.g. hips) to help the model resolve overlapping instances, or would that add more annotation noise than benefit given my image quality?

**Training so far:**
I've run two trainings with the bottom-up ID model (config screenshots attached). My latest results were:
- PCK@S: 0.5394
- mOKS: 0.7183
- mAP: 0.4413

I'd like to increase all of these, but mAP especially. I've noticed mOKS is reasonably high compared to mAP, which makes me think the issue is more about instance detection/identity association during overlap than keypoint localization itself.

**Hardware / environment:**
- Intel Core i9-13900F (2.00 GHz), 32 GB RAM
- NVIDIA GeForce RTX 4070 SUPER (12 GB)
- SLEAP v1.6.3, Python 3.13.13
- PyTorch backend (not TensorFlow)
- CUDA available: True

**Questions:**
1. Given the amount of overlap between animals, does it make sense to expand the skeleton, or should I focus on adding more labeled frames instead?
2. Are there specific config changes (backbone, filters, sigma, augmentation) that typically help mAP specifically in bottom-up ID setups with heavy occlusion?
3. My labeled set is only ~4% of my total frames (1,065 of 27,021). Should I substantially increase the number of labeled frames, and if so, is it better to prioritize quantity or to focus specifically on high-overlap/occlusion frames?

Attached below: a validation prediction example, my skeleton node list, a raw frame example, my training pipeline config, my model/data/augmentation config, and a labeled example frame.

Thanks in advance for any help!

<img width="1919" height="1028" alt="image" src="https://github.com/user-attachments/assets/2ef4dbfa-e4af-4e1f-ae7b-b006d7d6d750" />

<img width="421" height="350" alt="image" src="https://github.com/user-attachments/assets/13f86ae0-7e04-4a21-bb55-87975edde8b6" />

<img width="606" height="375" alt="image" src="https://github.com/user-attachments/assets/87cdb5fc-9601-4620-86e7-259eea5fae4c" />

<img width="972" height="794" alt="image" src="https://github.com/user-attachments/assets/23449603-45e7-4b76-b460-81cf649c3abd" />

<img width="984" height="924" alt="image" src="https://github.com/user-attachments/assets/5cbc04d5-a400-45f1-8624-dc60bab0e3d5" />

<img width="333" height="364" alt="image" src="https://github.com/user-attachments/assets/642e41ce-0ca8-4114-8c49-e77ecc878e01" />

<img width="631" height="576" alt="image" src="https://github.com/user-attachments/assets/081339c6-1edf-4816-b177-8ef9e44f847a" />

**Comment by gitttt-1234** [comment]
Hi @naaguilarr !

> I'd like to increase all of these, but mAP especially. I've noticed mOKS is reasonably high compared to mAP, which makes me think the issue is more about instance detection/identity association during overlap than keypoint localization itself.

OKS is a measure of how close the keypoints are between a GT and a predicted instance. Averaging that over all matched pairs (GT-prediction) gives the mean OKS. mAP is a hit/ miss metric, repeated at ten increasingly strict bars. We take OKS thresholds from 0.50 up to 0.95 and ask at each one, roughly, how many of your predictions count as correct then average those ten answers. (That's the COCO convention) So a raw score of 0.72 clears the 0.50–0.70 thresholds and then drops off: very few of them clear 0.80, 0.85, 0.90 or 0.95, and those contribute almost nothing to the average. That's why mAP is 0.44. 

I would recommend checking the **distance errors** in the metrics. If the 90%/95% percentiles are large, that's a localization issue. Also look at the **visibility metrics** to see how many GT nodes were predicted vs. missed. 

> Are there specific config changes (backbone, filters, sigma, augmentation) that typically help mAP specifically in bottom-up ID setups with heavy occlusion?

First thing to try would be increasing the **receptive field** (blue box in the training dialog window), it needs to be roughly as big as your animal. Right now it's **156 px**, which is smaller than one of your animals. Raise **Max Stride 32 → 64** (and filter rate to 1.5), which takes the receptive field to 316 px. If the keypoint localization is an issue, try setting sigma to 2.5 and output stride to 2 just for the first "Output stride" in the head config!

One thing worth checking, open `training_config.yaml` in your model's folder and find:

```yaml
model_config:
  head_configs:
    multi_class_bottomup:
      class_maps:
        classes:
```

It should list **exactly two** names for classes. If you see more than two, we might have to go back and look at the tracks assigned to the instance in the slp project.

> Given the amount of overlap between animals, does it make sense to expand the skeleton, or should I focus on adding more labeled frames instead? My labeled set is only ~4% of my total frames (1,065 of 27,021). Should I substantially increase the number of labeled frames, and if so, is it better to prioritize quantity or to focus specifically on high-overlap/occlusion frames?

The skeleton looks quite solid. On labels: having 1,065 across 12 videos is already a good set. If you want to focus on the overlapping ones: Generate some suggestion frames -> Run your current model on that -> Sort by score (check the suggestions tab -> Proofread a few of the bad ones esp. the high-overlap and mounting frames -> Retrain. Correcting 100 frames where the model is currently failing will help more than adding 500 easy ones. 

On the ID model generally: these can be tricky with similar-looking animals. I do see a mark on one of them, which I assume is the key to telling them apart, and from your predictions it does look like the model is managing it. Still, it'd be a worthwhile experiment to try a plain bottom-up model plus tracking instead. You might get some identity switches that way, but it would tell you whether the identity head is costing you pose accuracy worth knowing if the receptive field change doesn't move things enough.

Let us know if you have any questions!

Thanks,

Divya
